wagtail/core/actions/copy_for_translation.py

```diff
@@ -0,0 +1,152 @@
+from django.core.exceptions import PermissionDenied
+from django.db import transaction
+
+from wagtail.core.utils import find_available_slug
+
+
+class ParentNotTranslatedError(Exception):
+    """
+    Raised when a call to Page.copy_for_translation is made but the
+    parent page is not translated and copy_parents is False.
+    """
+
+    pass
+
+
+class CopyForTranslationPermissionError(PermissionDenied):
+    """
+    Raised when the page translation copy cannot be performed due to insufficient permissions.
+    """
+
+    pass
+
+
+class CopyForTranslationAction:
+    """
+    Creates a copy of this page in the specified locale.
+
+    The new page will be created in draft as a child of this page's translated
+    parent.
+
+    For example, if you are translating a blog post from English into French,
+    this method will look for the French version of the blog index and create
+    the French translation of the blog post under that.
+
+    If this page's parent is not translated into the locale, then a ``ParentNotTranslatedError``
+    is raised. You can circumvent this error by passing ``copy_parents=True`` which
+    copies any parents that are not translated yet.
+
+    The ``exclude_fields`` parameter can be used to set any fields to a blank value
+    in the copy.
+
+    Note that this method calls the ``.copy()`` method internally so any fields that
+    are excluded in ``.exclude_fields_in_copy`` will be excluded from the translation.
+    """
+
+    def __init__(
+        self,
+        page,
+        locale,
+        copy_parents=False,
+        alias=False,
+        exclude_fields=None,
+        user=None,
+        include_subtree=False,
+    ):
+        self.page = page
+        self.locale = locale
+        self.copy_parents = copy_parents
+        self.alias = alias
+        self.exclude_fields = exclude_fields
+        self.user = user
+        self.include_subtree = include_subtree
+
+    def check(self, skip_permission_checks=False):
+        # Permission checks
+        if (
+            self.user
+            and not skip_permission_checks
+            and not self.user.has_perms(["simple_translation.submit_translation"])
+        ):
+            raise CopyForTranslationPermissionError(
+                "You do not have permission to submit a translation for this page."
+            )
+
+    def walk(self, current_page):
+        for child_page in current_page.get_children():
+            action = CopyForTranslationAction(page=child_page, locale=self.locale)
+            action.execute(skip_permission_checks=True)
```

Comment:
I think we should avoid making actions construct and execute new actions. This would make it more consistent with with way ``CopyPageAction`` works and might be helpful if we later make these actions async.

We also are not currently passing down the ``user``/``alias`` to the sub-pages and I think these settings should be applied to the entire tree.

Reply:
Alright!